File: Task2.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <string>
#include <algorithm>
#include <iomanip>
#include <sstream>
#include <limits>
using namespace std;

struct Visitor {
    string ticket;
    int duration;
};

struct Window {
    int total_time = 0;
    vector<Visitor> visitors;
};

string generate_ticket() {
    static int counter = 1;
    stringstream ss;
    ss << "T" << setw(3) << setfill('0') << counter++;
    return ss.str();
}

vector<Window> distribute_queue(queue<Visitor>& q, int windows_count) {
    vector<Window> windows(windows_count);

    while (!q.empty()) {
        Visitor visitor = q.front();
        q.pop();

        auto min_window = min_element(windows.begin(), windows.end(),
            [](const Window& a, const Window& b) {
                return a.total_time < b.total_time;
            });

        min_window->visitors.push_back(visitor);
        min_window->total_time += visitor.duration;
    }
    return windows;
}

int main() {
    int windows_count = 0;
    string command;
    queue<Visitor> visitors_queue;
    
    while (true) {
        cout << ">>> Введите кол-во окон" << endl;
        cout << "<<< ";
        try {
            cin >> windows_count;
            if (cin.fail() || windows_count <= 0) {
                cin.clear();
                cin.ignore(numeric_limits<streamsize>::max(), '\n');
                throw invalid_argument("Неверный ввод. Введите положительное целое число");
            }
            break;
        } catch (const exception& e) {
            cerr << ">>> Ошибка: " << e.what() << endl;
        }
    }

    while (true) {
        cout << "<<< ";
        cin >> command;
        
        if (command == "ENQUEUE") {
            while (true) {
                int duration;
                try {
                    cin >> duration;
                    if (cin.fail() || duration <= 0) {
                        cin.clear();
                        cin.ignore(numeric_limits<streamsize>::max(), '\n');
                        throw invalid_argument("Неверный ввод. Введите положительное целое число");
                    }
                    
                    string ticket = generate_ticket();
                    visitors_queue.push({ticket, duration});
                    cout << ">>> " << ticket << endl;
                    break;
                } catch (const exception& e) {
                    cerr << ">>> Ошибка: " << e.what() << endl;
                }
            }
        } 
        else if (command == "DISTRIBUTE") {
            break;
        } 
        else {
            cout << ">>> Неизвестная команда. Используйте ENQUEUE или DISTRIBUTE" << endl;
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
        }
    }

    vector<Window> windows = distribute_queue(visitors_queue, windows_count);

    for (size_t i = 0; i < windows.size(); ++i) {
        cout << ">>> Окно " << i + 1 << " (" << windows[i].total_time << " минут): ";
        for (size_t j = 0; j < windows[i].visitors.size(); ++j) {
            if (j != 0) cout << ", ";
            cout << windows[i].visitors[j].ticket;
        }
        cout << endl;
    }
    return 0;
}

File: Task2.rs
use std::collections::VecDeque;
use std::io;

#[derive(Clone)]
struct Visitor {
    ticket: String,
    duration: i32,
}

struct Window {
    total_time: i32,
    visitors: Vec<Visitor>,
}

impl Window {
    fn new() -> Self {
        Window {
            total_time: 0,
            visitors: Vec::new(),
        }
    }
}

fn generate_ticket(counter: &mut i32) -> String {
    *counter += 1;
    format!("T{:03}", counter)
}

fn distribute_queue(q: &mut VecDeque<Visitor>, windows_count: usize) -> Vec<Window> {
    let mut windows = Vec::with_capacity(windows_count);
    for _ in 0..windows_count {
        windows.push(Window::new());
    }

    while let Some(visitor) = q.pop_front() {
        let duration = visitor.duration;  // Сохраняем duration до перемещения visitor
        let min_window = windows
            .iter_mut()
            .min_by_key(|w| w.total_time)
            .unwrap();

        min_window.visitors.push(visitor);
        min_window.total_time += duration;  // Используем сохраненное значение
    }
    windows
}

fn main() {
    let mut ticket_counter = 0;
    let mut visitors_queue = VecDeque::new();

    println!(">>> Введите кол-во окон");
    print!("<<< ");
    io::Write::flush(&mut io::stdout()).unwrap();

    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let windows_count: usize = input.trim().parse().unwrap();

    loop {
        print!("<<< ");
        io::Write::flush(&mut io::stdout()).unwrap();
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let parts: Vec<&str> = input.trim().split_whitespace().collect();

        if parts.is_empty() {
            continue;
        }

        match parts[0] {
            "ENQUEUE" => {
                if parts.len() < 2 {
                    println!("Необходимо указать длительность");
                    continue;
                }
                let duration: i32 = parts[1].parse().unwrap();
                let ticket = generate_ticket(&mut ticket_counter);
                visitors_queue.push_back(Visitor { ticket: ticket.clone(), duration });
                println!(">>> {}", ticket);
            }
            "DISTRIBUTE" => {
                break;
            }
            _ => {
                println!("Неизвестная команда");
            }
        }
    }

    let windows = distribute_queue(&mut visitors_queue, windows_count);

    for (i, window) in windows.iter().enumerate() {
        print!(">>> Окно {} ({} минут): ", i + 1, window.total_time);
        for (j, visitor) in window.visitors.iter().enumerate() {
            if j != 0 {
                print!(", ");
            }
            print!("{}", visitor.ticket);
        }
        println!();
    }
}
